components/variant-changer.js

```diff
@@ -0,0 +1,142 @@
+      this.updateProductPage();
+    }
+  }
+
+  handleSelectChange(event) {
+    const target = event.target;
+
+    if (target.tagName.toLowerCase() == "select") {
+      this.selectedOptions = Array.from(
+        this.querySelectorAll("select"),
+        (select) => select.value
+      );
+      this.currentVariant = this.getCurrentVariant();
+
+      this.urlReplace(this.currentVariant.id);
+
+      this.changeProductFormInputData(this.currentVariant.id);
+      this.updateProductPage();
+    }
+  }
+
+  setVariantFromUrl() {
+    const urlParams = new URLSearchParams(window.location.search);
+    const variantId = urlParams.get("variant");
+
+    if (variantId) {
+      this.changeProductFormInputData(variantId);
+    }
+  }
+
+  urlReplace(value) {
+    const currentUrlAttribute = "data-current-url";
+    const currentUrl = this.querySelector(
+      `[${currentUrlAttribute}]`
+    ).getAttribute(`${currentUrlAttribute}`);
+
+    window.history.replaceState({}, "", `${currentUrl}?variant=${value}`);
+  }
+
+  changeInputVariantToActive(target) {
+    const variantId = target.getAttribute(this.variantIdAttribute);
+    const variantGroup = this.querySelector(
+      `[${this.groupIdAttribute}="${variantId}"]`
+    );
+
+    variantGroup
+      .querySelectorAll(`[${this.variantTrigger}]`)
+      .forEach((variant) => {
+        variant.removeAttribute(this.variantActiveAttribute);
+      });
+
+    target.setAttribute(this.variantActiveAttribute, "");
+  }
+
+  changeProductFormInputData(variantId) {
+    const input = this.querySelector(`input[${this.formInputAttribute}]`);
+
+    input.value = variantId;
+  }
+
+  getSelectedOptions() {
+    return Array.from(
+      this.querySelectorAll(`[${this.variantActiveAttribute}]`)
+    ).map((option) => option.getAttribute(this.variantValueAttribute));
+  }
+
+  getCurrentVariant() {
+    const variants = JSON.parse(
+      this.querySelector('[type="application/json"]').innerHTML
+    );
+
+    return variants.find((variant) => {
+      return variant.options.every(
+        (value, index) => value === this.selectedOptions[index]
+      );
+    });
+  }
+
+  async updateProductPage() {
+    const submitButton = document.querySelector(
+      `[${this.submitButtonAttribute}]`
+    );
+    submitButton.disabled = true;
+
+    try {
+      const url = window.location.href;
+
+      const res = await fetch(url);
+
+      if (!res.ok) {
+        throw new Error(`Error: ${res.status} ${res.statusText}`);
+      }
+
+      const responseData = await res.text();
+      const resultsMarkup = new DOMParser().parseFromString(
+        responseData,
+        "text/html"
+      );
+
+      document.querySelector(`[${this.productSectionAttribute}]`).innerHTML =
+        resultsMarkup.querySelector(
+          `[${this.productSectionAttribute}]`
+        ).innerHTML;
+    } catch (error) {
+      console.error("Error: ", error);
+    } finally {
+      submitButton.disabled = false;
```

Comment:
I think we should remove this, when the DOM gets replaced, the state of the button should be determined by what Liquid returns as the starting state. You could be removing a disabled when it shouldn't be if the item was out of stock and Liquid determined that.